Viewport setup abstraction

Xottab-DUTY opened the pull request.

Xottab-DUTY committed:
Abstracted viewport setup under HW

SkyLoaderr commented on src/Layers/xrRender/R_Backend.h:
```diff
@@ -358,6 +358,7 @@ class ECORE_API CBackend
     void set_ClipPlanes(u32 _enable, Fplane* _planes = nullptr, u32 count = 0);
     void set_ClipPlanes(u32 _enable, Fmatrix* _xform = nullptr, u32 fmask = 0xff);
     IC void set_Scissor(Irect* rect = nullptr);
+    IC void SetViewport(D3D_VIEWPORT viewport) const;
```
I think it's not good to pass structure. Maybe use a reference of structure?

Xottab-DUTY replied:
I think it's even better to pass r-value reference.

vTurbine replied:
Yeah. And probably it’s time to rename ‘D3D_VIEWPORT’ into smth like ‘Viewport’

Xottab-DUTY replied:
I think it's a good name, which shows that DirectX is a head renderer and other backends implementations should mimic to DirectX behaviour.

Xottab-DUTY replied:
And personally I can't even imagine a good name for the type that would fit.

`Viewport` could be good, but...
```cpp
// public member variables should be PascalCase

struct ThatWasMadeForSomeReason
{
	SomeType SomeVar;
	... // more variables
	Viewport Viewport; // I'm not sure, but GCC won't be happy about this probably
                       // But even if it will, this looks bad.
}

Xottab-DUTY committed:
Removed wrapper around wrapper
Previously it was just wrapper, but after creating an abstraction in CBackend it became redundant

Xottab-DUTY committed:
SetViewport can now accept const reference or r-value reference.
Using r-value can be especially helpful for OpenGL

Xottab-DUTY committed:
Forgot to remove 'const' in the previous commit

vTurbine commented on src/Layers/xrRender/R_Backend.h:
```diff
@@ -358,6 +358,8 @@ class ECORE_API CBackend
     void set_ClipPlanes(u32 _enable, Fmatrix* _xform = nullptr, u32 fmask = 0xff);
     IC void set_Scissor(Irect* rect = nullptr);
+    IC void SetViewport(const D3D_VIEWPORT& viewport) const;
+    IC void SetViewport(D3D_VIEWPORT&& viewport) const;
```
It will be fine to have only the first implementation

Xottab-DUTY committed:
Removed redundant function overload

Xottab-DUTY committed:
Removed std::move

vTurbine reviewed (approved):
Looks good to me 👍
